### ISSUE TITLE
Combination of lwarmstart = .true. and iadv_(qt/thl) = 7 causes crash

### ISSUE THREAD

**Issue by schnierstein** [dalesteam/dales#68]
In `modstartup.f90`in `subroutine startup`:
```fortran
else !if lwarmstart

      call readrestartfiles
      ...
      call calc_halflev
      ...
      call baseprofs !call baseprofs

end if  ! end if (.not. warmstart)
```

If `iadv_(qt/thl) = 7` (Kappa Advection scheme) `calc_halflev` calls `halflev_kappa` from `advec_kappa.f90`

In `halflev_kappa` we divide by `rhobh` which is set to 0 before being read in in `baseprofs` only after the half-level calculation. The calculated `NaN`-value then happily propagates through the code and causes a segementation fault down the line.

I am now checking if the easy fix is just to put `call baseprofs` before `call calc_halflev`.

**Comment by fjansson** [comment]
Thank you for the report and analysis! I've only used kappa for (rain and passive) scalars. 

I'd welcome a fix for the startup sequence. But in general, is kappa useful for thl, qt? kappa guarantees positivity (no under/overshoots) at the price of being quite diffusive, if I understand it right. I think overshoots aren't a big issue for thl and qt, but extra diffusion there could be bad for the simulation. If noone needs kappa for thl, qt, I'd be in favor of removing support for it, because then we don't have to maintain halflev_kappa either.

The startup order can be tricky and fragile in my experience. Please let us know how the tests go.

**Comment by adoyenne** [comment]
Hello,

I also tested this with warmstart = .true. and idv = 7, since I simulate scalar tracers of co2. I agree that the reading of baseprofs should be before the calc_halflev, because if it calls after calc_halflev, the density is 0 in advec_kappa subroutine halflev_kappa(putin,putout). Thus, I set the sequence of calls here like this and it works correct now: 

 else !if lwarmstart

      call readrestartfiles
      call baseprofs ! call baseprofs before calc_halflev Arseni (probably should be here since the density is 0 in advec_kappa subroutine  halflev_kappa(putin,putout))
      
      um   = u0
      vm   = v0
      wm   = w0
      thlm = thl0
      qtm  = qt0
      svm  = sv0
      e12m = e120
      call calc_halflev
      exnf = (initial_presf/pref0)**(rd/cp)
      exnh = (initial_presh/pref0)**(rd/cp)

      do  j=2,j1
      do  i=2,i1
      do  k=2,k1
        thv0h(i,j,k) = (thl0h(i,j,k)+rlv*ql0h(i,j,k)/(cp*exnh(k))) &
                      *(1+(rv/rd-1)*qt0h(i,j,k)-rv/rd*ql0h(i,j,k))
      end do
      end do
      end do

      do  j=2,j1
      do  i=2,i1
      do  k=1,k1
        thv0(i,j,k) = (thl0(i,j,k)+rlv*ql0(i,j,k)/(cp*exnf(k))) &
                      *(1+(rv/rd-1)*qt0(i,j,k)-rv/rd*ql0(i,j,k))
      end do
      end do
      end do

      thvh=0.
      call slabsum(thvh,1,k1,thv0h,2-ih,i1+ih,2-jh,j1+jh,1,k1,2,i1,2,j1,1,k1) ! redefine halflevel thv using calculated thv
      thvh = thvh/ijtot

      thvf = 0.0
      call slabsum(thvf,1,k1,thv0,2-ih,i1+ih,2-jh,j1+jh,1,k1,2,i1,2,j1,1,k1)
      thvf = thvf/ijtot

      u0av = 0.0
      v0av = 0.0
      thl0av = 0.0
      th0av  = 0.0
      qt0av  = 0.0
      ql0av  = 0.0
      sv0av = 0.

      call slabsum(u0av  ,1,k1,u0  ,2-ih,i1+ih,2-jh,j1+jh,1,k1,2,i1,2,j1,1,k1)
      call slabsum(v0av  ,1,k1,v0  ,2-ih,i1+ih,2-jh,j1+jh,1,k1,2,i1,2,j1,1,k1)
      call slabsum(thl0av,1,k1,thl0,2-ih,i1+ih,2-jh,j1+jh,1,k1,2,i1,2,j1,1,k1)
      call slabsum(qt0av ,1,k1,qt0 ,2-ih,i1+ih,2-jh,j1+jh,1,k1,2,i1,2,j1,1,k1)
      call slabsum(ql0av ,1,k1,ql0 ,2-ih,i1+ih,2-jh,j1+jh,1,k1,2,i1,2,j1,1,k1)
      do n=1,nsv
        call slabsum(sv0av(1,n),1,k1,sv0(1,1,1,n),2-ih,i1+ih,2-jh,j1+jh,1,k1,2,i1,2,j1,1,k1)
      end do

      u0av  = u0av  /ijtot + cu
      v0av  = v0av  /ijtot + cv
      thl0av = thl0av/ijtot
      qt0av = qt0av /ijtot
      ql0av = ql0av /ijtot
      sv0av = sv0av /ijtot
      th0av  = thl0av + (rlv/cp)*ql0av/exnf
      thvh(1) = th0av(1)*(1+(rv/rd-1)*qt0av(1)-rv/rd*ql0av(1)) ! override first level

      do k=1,k1
        rhof(k) = presf(k)/(rd*thvf(k)*exnf(k))
      end do

      ! CvH - only do this for fixed timestepping. In adaptive dt comes from restartfile
      if(ladaptive .eqv. .false.) rdt=dtmax
      !call baseprofs !call baseprofs Arseni I put it above, after call readrestartfiles
      if(lopenbc) then
        call openboundary_readboundary
      endif

    end if  ! end if (.not. warmstart)

**Comment by fjansson** [comment]
@adoyenne good find! I think this explains the weird behavior you had on warm start.

Btw my opinion now is that kappa is nice to have for qt and thl advection, and that warm starts with kappa for those should be made to work. I think you're right about moving the calls, I just want to convince myself nothing else breaks.

**Comment by adoyenne** [comment]
@fjansson well, with fixing of this issue I just managed to simulate the warm start simulation without a crash at first time step due to segmentation fault (devision by 0). However, the problem with nudge lathernal boundaries in the warmstart simulation is still exist (the sharp increase of co2 at borders) and I still have not found what is the reason for it. Do you have some guess?

**Comment by fjansson** [comment]
Thank you @schnierstein and @adoyenne for reporting and fixing this, sorry I lost track of it.

Your call order fix looked good to me, so I committed that to the development branch to4.4.2_Fredrik. I also rebased 4.4_Ruisdael on top of that. Perhaps you now need to rebase your additions onto the new 4.4_Ruisdael branch.

@adoyenne perhaps the boundary problem is a separate issue, I don't know yet what it could be. It looks like something wrong happening only at the borders, apparently at the time of restart...
